From 618b6d95a9bb9c11cd8775de7ee78347462d4bbd Mon Sep 17 00:00:00 2001
From: myfluxi <linflux@arcor.de>
Date: Fri, 13 Dec 2013 18:15:30 +0100
Subject: [PATCH 484/507] msm: mdss: Use workqueue to send panel commands

Sending panel commands is known to be expensive, honor this by
scheduling this task. Also check for panel state and send commands
only when we're actually awake.

These fixes adress bug reports regarding behaviour during boot.

Signed-off-by: Simarpreet Singh <simar@linux.com>
---
 drivers/video/msm/mdss/mdss_dsi_panel.c |   28 ++++++++++++++++++++--------
 1 file changed, 20 insertions(+), 8 deletions(-)

diff --git a/drivers/video/msm/mdss/mdss_dsi_panel.c b/drivers/video/msm/mdss/mdss_dsi_panel.c
index 4e440d0..7498407 100644
--- a/drivers/video/msm/mdss/mdss_dsi_panel.c
+++ b/drivers/video/msm/mdss/mdss_dsi_panel.c
@@ -52,6 +52,7 @@ extern void lm3630_lcd_backlight_set_level(int level);
 
 static struct mdss_dsi_phy_ctrl phy_params;
 static struct mdss_panel_common_pdata *local_pdata;
+static struct work_struct send_cmds_work;
 struct mdss_panel_data *cmds_panel_data;
 
 void mdss_dsi_panel_pwm_cfg(struct mdss_dsi_ctrl_pdata *ctrl)
@@ -1203,18 +1204,13 @@ static int write_local_on_cmds(struct device *dev, const char *buf,
 	return rc;
 }
 
-/************************** sysfs interface ************************/
-
-static ssize_t write_kgamma_send(struct device *dev,
-				 struct device_attribute *attr,
-				 const char *buf, size_t count)
+static void send_local_on_cmds(struct work_struct *work)
 {
-	int rc;
 	struct mdss_dsi_ctrl_pdata *ctrl = NULL;
 
 	if (cmds_panel_data == NULL) {
 		pr_err("%s: Invalid input data\n", __func__);
-		return -EINVAL;
+		return;
 	}
 
 	ctrl = container_of(cmds_panel_data, struct mdss_dsi_ctrl_pdata,
@@ -1224,8 +1220,22 @@ static ssize_t write_kgamma_send(struct device *dev,
 		mdss_dsi_panel_cmds_send(ctrl, &local_pdata->on_cmds);
 
 	pr_info("%s\n", __func__);
+}
 
-	return rc;
+/************************** sysfs interface ************************/
+
+static ssize_t write_kgamma_send(struct device *dev,
+				 struct device_attribute *attr,
+				 const char *buf, size_t count)
+{
+	if (!cmds_panel_data->panel_info.panel_power_on) {
+		pr_err("%s: Panel off, failed to send commands\n", __func__);
+		return -EPERM;
+	}
+
+	schedule_work(&send_cmds_work);
+
+	return count;
 }
 
 static DEVICE_ATTR(kgamma_send, 0644, NULL, write_kgamma_send);
@@ -1319,6 +1329,8 @@ static int __devinit mdss_dsi_panel_probe(struct platform_device *pdev)
 	if (rc)
 		return rc;
 
+	INIT_WORK(&send_cmds_work, send_local_on_cmds);
+
 	local_pdata = &vendor_pdata;
 	if (!local_pdata)
 		return -EINVAL;
-- 
1.7.9.5

